gymnasium/dev_wrappers/time_aware_observation.py

```diff
@@ -0,0 +1,140 @@
+"""Wrapper for adding time aware observations to environment observation."""
+from collections import OrderedDict
+
+import jumpy as jp
+
+import gymnasium
+import gymnasium.spaces as spaces
+from gymnasium.core import ActType, ObsType
+from gymnasium.spaces import Box, Dict
+
+
+class TimeAwareObservationV0(gymnasium.ObservationWrapper):
+    """Augment the observation with time information of the episode.
+
+    Time can be represented as a normalized value between [0,1]
+    or by the number of timesteps remaining before truncation occurs.
+
+    Example:
+        >>> import gymnasium
+        >>> from gymnasium.wrappers import TimeAwareObservationV0
+        >>> env = gymnasium.make('CartPole-v1')
+        >>> env = TimeAwareObservationV0(env)
+        >>> env.observation_space
+        Dict(obs: Box([-4.8000002e+00 -3.4028235e+38 -4.1887903e-01 -3.4028235e+38], [4.8000002e+00 3.4028235e+38 4.1887903e-01 3.4028235e+38], (4,), float32), time: Box(0.0, 500, (1,), float32))
+        >>> _ = env.reset()
+        >>> env.step(env.action_space.sample())[0]
+        OrderedDict([('obs',
+        ...       array([ 0.02866629,  0.2310988 , -0.02614601, -0.2600732 ], dtype=float32)),
+        ...      ('time', array([0.002]))])
+
+    Flatten observation space example:
+        >>> env = gymnasium.make('CartPole-v1')
+        >>> env = TimeAwareObservationV0(env, flatten=True)
+        >>> env.observation_space
+        Box([-4.8000002e+00 -3.4028235e+38 -4.1887903e-01 -3.4028235e+38  0.0000000e+00], [4.8000002e+00 3.4028235e+38 4.1887903e-01 3.4028235e+38 500], (5,), float32)
+        >>> _ = env.reset()
+        >>> env.step(env.action_space.sample())[0]
+        array([-0.01232257,  0.19335455, -0.02244143, -0.32388705,  0.002 ], dtype=float32)
+    """
+
+    def __init__(self, env: gymnasium.Env, flatten=False, normalize_time=True):
+        """Initialize :class:`TimeAwareObservationV0`.
+
+        Args:
+            env: The environment to apply the wrapper
+            flatten: Flatten the observation to a `Box` of a single dimension
+            normalize_time: if `True` return time in the range [0,1]
+                otherwise return time as remaining timesteps before truncation
+        """
+        super().__init__(env)
+        self.flatten = flatten
+        self.normalize_time = normalize_time
+        self.max_timesteps = self._get_max_timesteps(env)
+        self.num_envs = getattr(env, "num_envs", 1)
+
+        self.time_aware_observation_space = Dict(
+            obs=env.observation_space, time=Box(0, self.max_timesteps, (self.num_envs,))
+        )
+
+        if self.flatten:
+            self.observation_space = spaces.flatten_space(
+                self.time_aware_observation_space
+            )
+        else:
+            self.observation_space = self.time_aware_observation_space
```

Comment:
Could we define the observation and _get_time_observation functions within init to avoid using the control flow every timestep?

Reply:
Updated in https://github.com/Farama-Foundation/Gymnasium/pull/22/commits/5202bf8b7c9e133b72a7eb2bbea8dea7b64ecb7a
This will avoid checking the control flow every time but maybe at the expense of some readability :thinking: are we good with that?

Reply:
refactored a little bit in https://github.com/Farama-Foundation/Gymnasium/pull/22/commits/e57935fe5bac47e51188e96b516eb0c110dd8d5c